lib/models/device.ts

```diff
@@ -3436,6 +3440,140 @@ const getDeviceModel = function (
+			 *
+			 *
+			 * @example
+			 * // get all device history entries between now - 20 days and now - 10 days
+			 * balena.models.device.history.getAllByDevice(999999, { fromDate: subDays(new Date(), 20), toDate: subDays(new Date(), 10)})
+			 */
+			async getAllByDevice(
+				uuidOrId: string | number,
+				{
+					fromDate,
+					toDate,
+					...options
+				}: PineOptions<DeviceHistory> & { fromDate?: Date; toDate?: Date } = {
+					fromDate: subDays(new Date(), defaultDaysOfRequestedHistory),
+				},
+			): Promise<DeviceHistory[]> {
+				let $filter = {
+					created_at: {
+						$ge: fromDate,
+						...(toDate ? { $le: toDate } : null),
+					},
+				};
+
+				if (isId(uuidOrId)) {
+					$filter = { ...$filter, ...{ tracks__device: uuidOrId } };
+				} else if (isUuid(uuidOrId)) {
+					$filter = { ...$filter, ...{ uuid: uuidOrId } };
+				} else {
+					throw new errors.BalenaInvalidParameterError('uuidOrId', uuidOrId);
+				}
+
+				return await pine.get<DeviceHistory>({
+					resource: 'device_history',
+					options: mergePineOptions(
+						{
+							$filter,
+						},
+						options,
+					),
+				});
+			},
+
+			/**
+			 * @summary Get all device history entries by application with time frame
+			 * @name getAllByApplication
+			 * @public
+			 * @function
+			 * @memberof balena.models.device.history
+			 *
+			 * @param {String|Number} slugOrUuidOrId - application slug (string), uuid (string) or id (number)
+			 * @param {Object} [options] - options
+			 * @param {Date} [options.fromDate=subDays(new Date(), 7)] - history entries older or equal to this date - default now() - 7 days
+			 * @param {Date} [options.toDate] - history entries younger or equal to this date
+			 * @fulfil {Object[]} - device history
+			 * @returns {Promise}
+			 *
+			 * @example
+			 * balena.models.device.history.getAllByApplication('myorganization/myapp').then(function(entries) {
+			 * 	console.log(entries);
+			 * });
+			 *
+			 * @example
+			 * balena.models.device.history.getAllByApplication(999999).then(function(entries) {
+			 * 	console.log(entries);
+			 * });
+			 *
+			 *  @example
+			 * // get all device history entries between now - 20 days and now - 10 days
+			 * balena.models.device.history.getAllByApplication(999999, { fromDate: subDays(new Date(), 20), toDate: subDays(new Date(), 10)})
+			 *
+			 */
+			async getAllByApplication(
+				slugOrUuidOrId: string | number,
+				{
+					fromDate,
+					toDate,
+					...options
+				}: PineOptions<DeviceHistory> & { fromDate?: Date; toDate?: Date } = {
+					fromDate: subDays(new Date(), defaultDaysOfRequestedHistory),
+				},
+			): Promise<DeviceHistory[]> {
+				const { id: applicationId } = await applicationModel().get(
+					slugOrUuidOrId,
+					{
+						$select: 'id',
+					},
+				);
+
+				return await pine.get<DeviceHistory>({
+					resource: 'device_history',
+					options: mergePineOptions(
+						{
+							$filter: {
+								created_at: {
+									$ge: fromDate,
```

Comment:
I think that you still need to check whether `fromDate` has a value or not here since atm it can be null.